frontend/catalyst/pennylane_extensions.py

```diff
@@ -921,6 +922,105 @@ def check_is_iterable(x, hint):
     return results
 
 
+class Zne:
+    """An object that specifies how a circuit is mitigated with ZNE.
+
+    Args:
+        fn (Callable): the circuit to be mitigated with ZNE.
+        scale_factors (array[int]): the range of noise scale factors used.
+        deg (int): the degree of the polymonial used for fitting.
+
+    Raises:
+        TypeError: Non-QNode object was passed as `fn`.
+    """
+
+    def __init__(self, fn: Callable, scale_factors: jax.numpy.ndarray, deg: int):
+        if not isinstance(fn, qml.QNode):
+            raise TypeError(f"A QNode is expected, got the classical function {fn}")
+        self.fn = fn
+        self.__name__ = f"zne.{getattr(fn, '__name__', 'unknown')}"
+        self.scale_factors = scale_factors
+        self.deg = deg
+
+    def __call__(self, *args, **kwargs):
+        """Specifies the an actual call to the folded circuit."""
+        jaxpr = jaxpr = jax.make_jaxpr(self.fn)(*args)
+        shapes = [out_val.shape for out_val in jaxpr.out_avals]
+        dtypes = [out_val.dtype for out_val in jaxpr.out_avals]
+        if any(shapes):
+            raise TypeError("Only expectations values and classical scalar values can be returned.")
+        if len(set(dtypes)) != 1:
+            raise TypeError(
+                "Dtypes of expectation values and classical classical values must match."
+            )
```

Comment:
`dtypes` here should probably always be a floating-point. I compiled a qnode returning an integer and got the following error message:

```
mitigated_qnode:4:10: error: 'mitigation.zne' op result #0 must be floating-point or ranked tensor of floating-point values, but got 'tensor<3xi64>'
    %1 = "mitigation.zne"(%arg0, %0) {callee = @circuit} : (tensor<f64>, tensor<3xi64>) -> tensor<3xi64>
         ^
mitigated_qnode:4:10: note: see current operation: %1 = "mitigation.zne"(%arg0, %0) {callee = @circuit} : (tensor<f64>, tensor<3xi64>) -> tensor<3xi64>
Failed to parse module as MLIR source, retrying parsing as LLVM source
mitigated_qnode: mitigated_qnode:1:8: error: expected 'module asm'
module @mitigated_qnode {
       ^
Failed to parse module as LLVM source
```

Reply:
I have added a check for the float kind